Apps/FlightSoftware/Watchdog2/src/i2c.c

```diff
@@ -0,0 +1,318 @@
+    UCB0CTLW0 |= UCMODE_3 | UCMST | UCSSEL__SMCLK | UCSYNC; // I2C master mode, SMCLK
+    UCB0BRW = 160;                            // fSCL = SMCLK/160 = ~100kHz
+    UCB0I2CSA = I2C_SLAVE_ADDR;                   // Slave Address
+    UCB0CTLW0 &= ~UCSWRST;                    // Clear SW reset, resume operation
+    UCB0IE |= UCNACKIE;
+}
+
+
+
+/* For slave device with dev_addr, read the data specified in slaves reg_addr.
+ * The received data is available in ReceiveBuffer
+ *
+ * dev_addr: The slave device address.
+ *           Example: I2C_SLAVE_ADDR
+ * reg_addr: The register or command to send to the slave.
+ *           Example: CMD_TYPE_0_SLAVE
+ * count: The length of data to read
+ *           Example: TYPE_0_LENGTH
+ *  */
+I2C_Mode I2C_Master_ReadReg(uint8_t dev_addr, uint8_t reg_addr, uint8_t count)
+{
+    /* Initialize state machine */
+    MasterMode = TX_REG_ADDRESS_MODE;
+    TransmitRegAddr = reg_addr;
+    RXByteCtr = count;
+    TXByteCtr = 0;
+    ReceiveIndex = 0;
+    TransmitIndex = 0;
+
+    /* Initialize slave address and interrupts */
+    UCB0I2CSA = dev_addr;
+    UCB0IFG &= ~(UCTXIFG + UCRXIFG);       // Clear any pending interrupts
+    UCB0IE &= ~UCRXIE;                       // Disable RX interrupt
+    UCB0IE |= UCTXIE;                        // Enable TX interrupt
+
+    UCB0CTLW0 |= UCTR + UCTXSTT;             // I2C TX, start condition
+
+    __bis_SR_register(GIE);              // Enable interrupts
+    __delay_cycles(10000);               // give fuel gauge time to respond
+    return MasterMode;
+
+}
+
+/* For slave device with dev_addr, writes the data specified in *reg_data
+ *
+ * dev_addr: The slave device address.
+ *           Example: I2C_SLAVE_ADDR
+ * reg_addr: The register or command to send to the slave.
+ *           Example: CMD_TYPE_0_MASTER
+ * *reg_data: The buffer to write
+ *           Example: MasterType0
+ * count: The length of *reg_data
+ *           Example: TYPE_0_LENGTH
+ *  */
+I2C_Mode I2C_Master_WriteReg(uint8_t dev_addr, uint8_t reg_addr, uint8_t *reg_data, uint8_t count)
+{
+    /* Initialize state machine */
+    MasterMode = TX_REG_ADDRESS_MODE;
+    TransmitRegAddr = reg_addr;
+
+    //Copy register data to TransmitBuffer
+    CopyArray(reg_data, TransmitBuffer, count);
+
+    TXByteCtr = count;
+    RXByteCtr = 0;
+    ReceiveIndex = 0;
+    TransmitIndex = 0;
+
+    /* Initialize slave address and interrupts */
+    UCB0I2CSA = dev_addr;
+    UCB0IFG &= ~(UCTXIFG + UCRXIFG);       // Clear any pending interrupts
+    UCB0IE &= ~UCRXIE;                       // Disable RX interrupt
+    UCB0IE |= UCTXIE;                        // Enable TX interrupt
+
+    UCB0CTLW0 |= UCTR + UCTXSTT;             // I2C TX, start condition
+
+    __bis_SR_register(GIE);              // Enable interrupts
+    __delay_cycles(10000);               // give fuel gauge some time to respond
+    return MasterMode;
+}
+
+void CopyArray(uint8_t *source, uint8_t *dest, uint8_t count)
```

Comment:
Is `memcpy` not available?

Reply:
@SchwartzCode

Reply:
Per [MSP430 Optimizing C/C++ Compiler v20.12.0.STS User's Guide](https://www.ti.com/lit/ug/slau132x/slau132x.pdf?ts=1620830388518) page 149, `memcpy()` is available in `string.h` and `bcopy()` (equivalent to `memmove()`) is defined in `strings.h`. Either of these could be used in place of this function. Obviously this guidance may be incorrect if we're using a different version of the compiler.